File: MinecraftManhunt/src/main/java/com/yoonicode/minecraftmanhuntplus/PluginMain.java
package com.yoonicode.minecraftmanhuntplus;

import com.yoonicode.minecraftmanhuntplus.respawn_inventory.InventoryGenerator;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.World;
import org.bukkit.entity.Player;
import org.bukkit.plugin.java.JavaPlugin;
import org.bukkit.scoreboard.Scoreboard;
import org.bukkit.scoreboard.ScoreboardManager;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.logging.Logger;

import static org.bukkit.Bukkit.getScoreboardManager;

public class PluginMain extends JavaPlugin {
    public ArrayList<Player> hunters = new ArrayList<Player>();
    public ArrayList<Player> runners = new ArrayList<Player>();
    public ArrayList<Player> spectators = new ArrayList<Player>();
    public HashMap<String, String> targets = new HashMap<String, String>();
    public HashMap<String, Location> portals = new HashMap<String, Location>();
    public Logger logger;
    public World world;
    public PluginCommands commands;
    public boolean debugMode = false;
    boolean compassEnabledInNether;
    public HashMap<Player, Integer> hunterDeaths = new HashMap<>();
    public HashMap<Player, Integer> runnerDeaths = new HashMap<>();
    private TaskManager taskManager = new TaskManager(this);
    private int gameState = 0;
    private InventoryGenerator itemGenerator;
    private boolean isPaused;
    private boolean gameIsOver;

    public boolean playerIsOnTeam(Player player){
        return hunters.stream().anyMatch(member->member.getName().equals(player.getName()))
                || runners.stream().anyMatch(member->member.getName().equals(player.getName()))
                || spectators.stream().anyMatch(member->member.getName().equals(player.getName()));
    }

    @Override
    public void onEnable() {
        logger = Logger.getLogger("com.yoonicode.minecraftmanhuntplus.PluginMain");
        logger = getLogger();
        logger.info("Minecraft Manhunt plugin enabled!");
        List<World> worlds = Bukkit.getWorlds();
        if(worlds.size() < 1){
            logger.warning("Could not detect main world! Plugin will not work.");
        }
        world = worlds.get(0);
        saveDefaultConfig();
        debugMode = getConfig().getBoolean("debugMode", false);
        compassEnabledInNether = getConfig().getBoolean("compassEnabledInNether", true);
        getServer().getPluginManager().registerEvents(new PluginListener(this), this);
        commands = new PluginCommands(this);
        for(String command : PluginCommands.registeredCommands){
            this.getCommand(command).setExecutor(commands);
        }
        this.itemGenerator = new InventoryGenerator(this);
        ScoreboardManager scoreboardManager = getScoreboardManager();
        Scoreboard board = scoreboardManager.getMainScoreboard();
        Bukkit.getServer().dispatchCommand(Bukkit.getConsoleSender(), "weather clear 199999999");
        PauseHandler.onEnable(this);
    }

    public World getWorld() {
        return world;
    }


    @Override
    public void onDisable() {
        logger.info("Minecraft Manhunt plugin disabled!");
        PauseHandler.onDisable(this);
    }

    public TaskManager getTaskManager() {
        return taskManager;
    }

    public int getGameState() {
        return gameState;
    }

    public void setGameState(int gameState) {
        this.gameState = gameState;
    }

    public void incrementGameState(){
        this.gameState++;
        PauseHandler.updateGame(this);
    }

    public InventoryGenerator getItemGenerator() {
        return itemGenerator;
    }

    public boolean isPaused() {
        return isPaused;
    }

    public void setPaused(boolean paused) {
        isPaused = paused;
    }

    public boolean isGameIsOver() {
        return gameIsOver;
    }

    public void setGameIsOver(boolean gameIsOver) {
        this.gameIsOver = gameIsOver;
        getConfig().set("gameOver",true);
        saveConfig();
    }
}

File: MinecraftManhunt/src/main/java/com/yoonicode/minecraftmanhuntplus/PluginMain.kt
package com.yoonicode.minecraftmanhuntplus

import com.yoonicode.minecraftmanhuntplus.SessionHandler.onEnable
import com.yoonicode.minecraftmanhuntplus.SessionHandler.onDisable
import org.bukkit.plugin.java.JavaPlugin
import org.bukkit.entity.Player
import java.util.HashMap
import org.bukkit.World
import com.yoonicode.minecraftmanhuntplus.respawn_inventory.InventoryGenerator
import com.yoonicode.minecraftmanhuntplus.game_state.GameStateCalculator
import org.bukkit.Bukkit
import org.bukkit.Location
import java.util.logging.Logger

class PluginMain : JavaPlugin() {
    @JvmField
    var hunters = mutableListOf<Player>()
    @JvmField
    var runners = mutableListOf<Player>()
    var spectators = mutableListOf<Player>()
    @JvmField
    var targets = HashMap<String, String>()
    @JvmField
    var portals = HashMap<String, Location>()
    var world: World? = null;
    var commands: PluginCommands = PluginCommands(this)
    var debugMode = false

    var mainLogger: Logger = Logger.getLogger("com.yoonicode.minecraftmanhuntplus.PluginMain")

    @JvmField
    var compassEnabledInNether = false
    var hunterDeaths = HashMap<Player, Int>()
    var runnerDeaths = HashMap<Player, Int>()
    val taskManager = TaskManager(this)
    var gameState = 0.0
    var itemGenerator: InventoryGenerator = InventoryGenerator(this)

    var isPaused = false
    var isGameIsOver = false
        set(gameIsOver) {
            field = gameIsOver
            config["gameOver"] = true
            saveConfig()
        }
    var gameStateCalculator: GameStateCalculator = GameStateCalculator(this)

    fun playerIsOnTeam(player: Player): Boolean {
        return (hunters.stream().anyMatch { member: Player -> member.name == player.name }
                || runners.stream().anyMatch { member: Player -> member.name == player.name }
                || spectators.stream().anyMatch { member: Player -> member.name == player.name })
    }

    /**
     * When the plugin is enabled (when the server is started)
     */
    override fun onEnable() {
        mainLogger.info("Minecraft Manhunt plugin enabled!")
        val worlds = Bukkit.getWorlds()
        if (worlds.size < 1) {
            mainLogger.warning("Could not detect main world! Plugin will not work.")
        }
        world = worlds[0]
        saveDefaultConfig()
        debugMode = config.getBoolean("debugMode", false)
        compassEnabledInNether = config.getBoolean("compassEnabledInNether", true)
        server.pluginManager.registerEvents(PluginListener(this), this)
        commands = PluginCommands(this)
        for (command in PluginCommands.registeredCommands) {
            getCommand(command)?.setExecutor(commands)
        }
        onEnable(this)
    }

    /**
     * When the plugin is disabled (when the server is closed)
     */
    override fun onDisable() {
        mainLogger.info("Minecraft Manhunt plugin disabled!")
        onDisable(this)
    }

}